clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -5791,12 +5783,32 @@ void SemaSYCL::MarkDevices() {
   }
 }
 
+static bool CheckFreeFunctionDiagnostics(Sema &S, FunctionDecl *FD) {
+  if (FD->isVariadic()) {
+    S.Diag(FD->getLocation(), diag::err_free_function_variadic_args);
+    return false;
+  }
+
+  if (!FD->getReturnType()->isVoidType()) {
+    S.Diag(FD->getLocation(), diag::err_free_function_return_type);
+    return false;
+  }
+
+  bool ReturnStatus = true;
+  for (ParmVarDecl *Param : FD->parameters()) {
+    if (Param->hasDefaultArg()) {
+      S.Diag(Param->getLocation(), diag::err_free_function_with_default_arg)
+          << Param->getSourceRange();
+      ReturnStatus = false;
+    }
+  }
+  return ReturnStatus;
```

Comment:
LLVM overall prefers early exits
```suggestion
  for (ParmVarDecl *Param : FD->parameters()) {
    if (Param->hasDefaultArg())
      return S.Diag(Param->getLocation(), diag::err_free_function_with_default_arg)
          << Param->getSourceRange();
  }
  return false;
```